controllers/configurationpolicy_controller.go

```diff
@@ -1158,81 +1154,99 @@ func (r *ConfigurationPolicyReconciler) handleObjectTemplates(plc policyv1.Confi
 				"desiredName", templateObjs[indx].name,
 				"index", indx,
 			)
-			var names []string
-			var compliant bool
-			var reason, objKind string
-			var related []policyv1.RelatedObject
-
-			if mapping != nil {
-				names, compliant, reason, objKind, related, statusUpdateNeeded = r.handleObjects(
-					objectT, ns, templateObjs[indx], indx, &plc, mapping, unstruct,
-				)
-			}
 
-			if statusUpdateNeeded {
-				parentStatusUpdateNeeded = true
-			}
+			if mappingErrResult != nil {
+				nsToResults[ns] = *mappingErrResult
 
-			if objKind != "" {
-				// object template enforced, objKind is empty
-				kind = objKind
+				continue
 			}
 
-			if mergeMessageEnforce {
-				names = []string{templateObjs[indx].name}
-			}
+			related, result := r.handleObjects(objectT, ns, templateObjs[indx], indx, &plc, mapping, unstruct)
 
-			// object template enforced, already handled in handleObjects
-			if names == nil {
-				handled = true
-			} else {
-				// when multiple messages and enforce mode, it doesn't mean remediactionAction is changed
-				enforce = false
+			nsToResults[ns] = result
+
+			for _, object := range related {
+				relatedObjects = updateRelatedObjectsStatus(relatedObjects, object)
 			}
+		}
 
-			// violations for enforce configurationpolicies are already handled in handleObjects,
-			// so we only need to generate a violation if the remediationAction is set to inform
-			// Or multiple namespaces and enforce use this
-			if !handled && !enforce {
-				if !compliant {
-					if len(names) == 0 {
-						numNonCompliant++
-					} else {
-						numNonCompliant += len(names)
-					}
+		// Each index is a batch of compliance events to be set on the ConfigurationPolicy before going on to the
+		// next one. For example, if an object didn't match and was enforced, there would be an event that it didn't
+		// match in the first batch, and then the second batch would be that it was updated successfully.
+		eventBatches := []map[string]*objectTmplEvalResultWithEvent{}
 
-					nonCompliantObjects[ns] = map[string]interface{}{
-						"names":  names,
-						"reason": reason,
-					}
-				} else {
-					numCompliant += len(names)
-					compliantObjects[ns] = map[string]interface{}{
-						"names":  names,
-						"reason": reason,
-					}
-				}
+		for ns, r := range nsToResults {
+			for len(eventBatches) < len(r.events) {
```

Comment:
I am lost here. eventBatches init with  `[]map[string]*objectTmplEvalResultWithEvent{}`, then append into []map[string]*objectTmplEvalResultWithEvent{{}},... is there any chance to break from this loop?